### ISSUE TITLE
powershell_script user attribute: no method for uid

### ISSUE THREAD

**Issue by docwhat** [chef/chef#2850]
This is copied from [CHEF-5344](https://tickets.opscode.com/browse/CHEF-5344):

> using the `powershell_script` resource and documentation states also has the `user` attribute like execute or bash on linux.
> The `powershell_script` resource executes the script fine, unless the `user` attribute is supplied. I used a local user account which chef created for testing purposes.
> I saw a similar ticket for the powershell cookbook that julian dunn closed recently, but nothing for chef

This is still a problem in Chef-Client 12.0.3

It seems `Mixlib::ShellOut` has the ability to change user (requiring a password, and possibly a domain), but the `powershell_script` resource/provider hasn't been updated.

**Comment by docwhat** [comment]
I searched, but couldn't find the ticket from the powershell cookbook mentioned.

**Comment by alex-williams** [comment]
I've hit this too and it's standing in the way of some important stuff. I've got a different issue with joining a Windows 2012 domain, because SYSTEM on 2012 does not appear to be allowed to perform the action of joining the domain. This was the workaround, to use the user feature in powershell_script, to run the commands necessary for joining the domain, in a different script, as a different user.

**Comment by docwhat** [comment]
If you use the `chef-client::task` instead of the windows service you can have it run as Administrator. As a work around.

**Comment by alex-williams** [comment]
I'll give that a try if my current test fails. Thanks @docwhat

**Comment by smurawski** [comment]
@alex-williams The domain join problem is mostly unrelated.  The only way domain join works under the local Administrator account is when the local Administrator password is the same as the domain's Administrator password and NTLM auth just matches that up.  No local user account (or service account) has rights to join a machine to a domain.  Only domain accounts with rights to join a computer account can do so.  In a default configuration, active directory will let a normal user account join 10 machines to the domain.  Most production domains I've worked on disable that and require specific accounts to be used to join machines to domains.

**Comment by smurawski** [comment]
@docwhat Agreed this is a bug and should get some attention. Paging @btm, @jdmundrawala, @ksubrama :)  Likely won't make it for the 12.1.0 release, but we'll get some attention on it.

**Comment by alex-williams** [comment]
@smurawski I think you've misunderstood what I was trying to do. I wasn't attempting to join the domain as the local admin, I was attempting to run the powershell script within the cookbook, as a user other than SYSTEM. It appeared at the time, that there was an issue running the Add-Computer command as SYSTEM on 2012 server, so I was attempting to run it as Administrator, with 'user'. As it turned out, the original issue wasn't an issue and was resolved another way, but this is still a feature that would be very desirable.

**Comment by lamont-granquist** [comment]
So if powershell_script calls script which calls execute, then this is likely a bug that should be addressed by teaching the execute resource how to properly read the `user` property off itself on windows, which would fix all the windows script resources (i.e. this is also likely a problem with batch and if anyone expects bash and such to work under cygwin, etc).

**Comment by lamont-granquist** [comment]
I suspect this is easy as long as its just wiring up the execute resource to working code in mixlib-shellout it should be a couple lines of code plus tests...

**Comment by adamedx** [comment]
Yes, this has never been implemented, and as @lamont-granquist suggests, it may be simply a matter of integrating the resource with mixlib-shellout's support for running as an alternate user.

**Comment by rayterrill** [comment]
Is this still a problem? I have exactly this problem - I need to run a command as a domain user. I see a note about powershell_out - is that a supported/usable way to go?

**Comment by lamont-granquist** [comment]
powershell_out is a stable API that can be used

**Comment by rayterrill** [comment]
What's up with this open issue, then? Does this mean it's something that will be introduced into powershell_script at some point (as in I should tag any code I do on my end to update it to use powershell_script when that's added), or should I go ahead and use powershell_out and be done with it?

**Comment by rayterrill** [comment]
Does anyone have a powershell_out example using a different set of credentials (i.e. chef-client is running as Local System, but I want the PowerShell command to run as a domain account)? I can't find anything like that, or maybe I'm misunderstanding how this works.

**Comment by btm** [comment]
@rayterrill we'll look at testing and adding the user/password properties to `powershell_script` soon

`powershell_out` takes an options hash as a second argument. Specify a username and password in the hash like so:
```
powershell_out(command_to_run, { "user": username, "password": password})
```

It looks like you can pass `domain` too. all these just pass through to powershell_out.

**Comment by btm** [comment]
This should be fixed when https://github.com/chef/chef/pull/5722 is complete.

**Comment by btm** [comment]
We've split the relevant portion of https://github.com/chef/chef/pull/5722 that fixes this issue out into https://github.com/chef/chef/pull/5764, which is ready for review.

**Comment by btm** [comment]
This is fixed in 12.19.8+ (only available in current at this time) by passing `user`, `password`, and `domain` to to `powershell_script` (and other execute resources should work too).
